### ISSUE TITLE
macos: GRPC assertion failures and hangs occasionally show up in CI builds

### ISSUE THREAD

**Issue by byllyfish** [byllyfish/finsy#306]
1. GRPC assertion in tests.  (grpcio-1.58.0, protobuf-4.24.3, py-3.10/macos-latest) (Action 826)

```
E0928 20:25:17.313613000 4560561664 sync.cc:57]                        ASSERTION FAILED: pthread_mutex_lock(mu) == 0
Fatal Python error: Aborted
```

2. When running code coverage, a "Waiting" message that persists until timeout. (grpcio-1.58.0, protobuf-4.24.3, py-3.11/macos-latest) (Action 818)

```
E0917 20:51:29.169858000 4742030848 original_thread_pool.cc:232]       Waiting for thread pool to idle before forking (1 to 0)
E0917 20:51:32.298273000 4742030848 original_thread_pool.cc:232]       Waiting for thread pool to idle before forking (1 to 0)
E0917 20:51:35.325750000 4742030848 original_thread_pool.cc:232]       Waiting for thread pool to idle before forking (1 to 0)
E0917 20:51:38.401483000 4742030848 original_thread_pool.cc:232]       Waiting for thread pool to idle before forking (1 to 0)
... (continues until github action times out)
```

3. GRPC assertion in tests.  (grpcio-1.57.0, protobuf-4.24.2, py-3.11/macos-latest) (Action 810)

```
E0911 01:50:00.267984000 4360189440 sync.cc:57]                        ASSERTION FAILED: pthread_mutex_lock(mu) == 0
Fatal Python error: Aborted
```

**Comment by byllyfish** [comment]
Still happening  on Mac OS latest (Action 875).

```
E1027 00:38:03.293649000 4571739648 sync.cc:57]                        ASSERTION FAILED: pthread_mutex_lock(mu) == 0
tests/test_p4schema.py::test_p4info_repr[p4info_file1] 
Fatal Python error: Aborted

Current thread 0x00000001107f3600 (most recent call first):
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/subprocess.py", line 1796 in _execute_child
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/subprocess.py", line 971 in __init__
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/subprocess.py", line 503 in run
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/subprocess.py", line 421 in check_output
  File "/Users/runner/work/finsy/finsy/tests/test_p4schema.py", line 190 in _format_source_code
  File "/Users/runner/work/finsy/finsy/tests/test_p4schema.py", line 174 in test_p4info_repr
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/python.py", line 194 in pytest_pyfunc_call
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/python.py", line 1792 in runtest
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/runner.py", line 169 in pytest_runtest_call
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/runner.py", line 262 in <lambda>
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/runner.py", line 341 in from_call
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/runner.py", line 261 in call_runtest_hook
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/runner.py", line 222 in call_and_report
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/runner.py", line 133 in runtestprotocol
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/runner.py", line 114 in pytest_runtest_protocol
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/main.py", line 350 in pytest_runtestloop
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/main.py", line 325 in _main
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/main.py", line 271 in wrap_session
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/main.py", line 318 in pytest_cmdline_main
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/config/__init__.py", line 169 in main
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/config/__init__.py", line 192 in console_main
  File "/Users/runner/hostedtoolcache/Python/3.10.13/x64/bin/pytest", line 8 in <module>

Extension modules: grpc._cython.cygrpc, google._upb._message (total: 2)
/Users/runner/work/_temp/98a5cf71-cea4-408f-864a-89471f3ef487.sh: line 1:  1576 Abort trap: 6           pytest
Error: Process completed with exit code 134.
```

**Comment by byllyfish** [comment]
I'm still seeing rare CI failures on macos-latest. Always with Python 3.10, never a newer version. The GHA is cancelled after running for > 9 minutes. Unlike the above logs, there is no other debug output or assertion failure; it's just a hang but in the same spot as above.

```
tests/test_p4schema.py::test_p4info_repr[p4info_file0] PASSED
Error: The operation was canceled.
```

**Comment by byllyfish** [comment]
The test_p4info_repr tests were invoking the `black` formatter as a subprocess. On macos, this was occasionally triggering a hang in the CI system that I could never reproduce locally. The hang was related to a custom threadpool spawned by the grpc server, I think, reacting to the process being forked.

I changed the test code to invoke the `black` formatter programmatically, as a library call. No more forking a subprocess. I have not seen any hangs since.

Closing this issue.
